nat-lab/tests/test_batching.py

```diff
@@ -0,0 +1,304 @@
+                ["stun", "local", "upnp"],
+                BatchTestFeatures(
+                    direct_interval=4,
+                    direct_threshold=0,
+                    is_meshnet=True,
+                    ip_stack=IPStack.IPv4,
+                ),
+                NO_BATCH_DOCKER_CONE_CLIENT_2_histogram_check,
+            ),
+            (
+                ConnectionTag.DOCKER_SHARED_CLIENT_1,
+                ["stun", "local", "upnp"],
+                BatchTestFeatures(
+                    direct_interval=4,
+                    direct_threshold=0,
+                    is_meshnet=False,
+                    ip_stack=IPStack.IPv4,
+                ),
+                NO_BATCH_DOCKER_SHARED_CLIENT_1_histogram_check,
+            ),
+        ])
+    ),
+]
+
+
+def BATCHED_DOCKER_CONE_CLIENT_1_histogram_check(hs: HistogramResult):
+    assert hs.outgoing[0] == 6
+    assert sum(hs.outgoing[7:8]) == 4
+    assert sum(hs.outgoing[1:7]) == 0
+
+
+def BATCHED_DOCKER_CONE_CLIENT_2_histogram_check(hs: HistogramResult):
+    assert hs.outgoing[0] == 6
+    assert sum(hs.outgoing[8:9]) == 4
+    assert sum(hs.outgoing[1:7]) == 0
+
+
+def BATCHED_DOCKER_SHARED_CLIENT_1_histogram_check(hs: HistogramResult):
+    assert hs.outgoing[0] == 6
+    assert sum(hs.outgoing[8:9]) == 4
+    assert sum(hs.outgoing[1:7]) == 0
+
+
+CLIENTS_BATCHING = [
+    pytest.param(
+        _generate_setup_parameters([
+            (
+                ConnectionTag.DOCKER_CONE_CLIENT_1,
+                ["stun", "local", "upnp"],
+                BatchTestFeatures(
+                    direct_interval=4,
+                    direct_threshold=2,
+                    is_meshnet=True,
+                    ip_stack=IPStack.IPv4,
+                ),
+                BATCHED_DOCKER_CONE_CLIENT_1_histogram_check,
+            ),
+            (
+                ConnectionTag.DOCKER_CONE_CLIENT_2,
+                ["stun", "local", "upnp"],
+                BatchTestFeatures(
+                    direct_interval=4,
+                    direct_threshold=2,
+                    is_meshnet=True,
+                    ip_stack=IPStack.IPv4,
+                ),
+                BATCHED_DOCKER_CONE_CLIENT_2_histogram_check,
+            ),
+            (
+                ConnectionTag.DOCKER_SHARED_CLIENT_1,
+                ["stun", "local", "upnp"],
+                BatchTestFeatures(
+                    direct_interval=4,
+                    direct_threshold=2,
+                    is_meshnet=False,
+                    ip_stack=IPStack.IPv4,
+                ),
+                BATCHED_DOCKER_SHARED_CLIENT_1_histogram_check,
+            ),
+        ])
+    ),
+]
+
+
+async def get_batch_histogram(
+    target: TargetMachine, mesh_node: Node
+) -> HistogramResult:
+    print(f"Running batch observer for target: {target.tag}")
+
+    await asyncio.sleep(2)
+    async with BatchObserver(target).run() as bo:
+        await asyncio.sleep(20)
```

Comment:
Can this sleep be avoided. Maybe we know the expected count of ping packets and could convert this to a loop that keeps getting the pcap and only breaks when the expected packet count have been found?

Reply:
Hmm. Didn't think about the loop idea to be honest, it sounds pretty nice, I will consider it and see how it goes :thinking: